magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

KrasnoshchokBohdan opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
As the author of the issue mentioned, if we add a special price using POST <host>/rest/<store_code>/V1/products/special-price specifying in the request body date like "price_to": "2017-07-16 23:59:59" i.e. time including hours, minutes and seconds, then as a result we will get that the special price will be valid for one day longer than the specified date

this happens in the method \Magento\Framework\Stdlib\DateTime\Timezone::isScopeDateInInterval where there is code that literally adds +1 day

```
      if ($dateTo) {
            // fix date YYYY-MM-DD 00:00:00 to YYYY-MM-DD 23:59:59
            $toTimeStamp += 86400;
        }
```
according to the example from the [documentation](https://developer.adobe.com/commerce/webapi/rest/modules/catalog/catalog-pricing/) (where "price_to": "2017-07-16 23:59:59" including hours is specified) we get this execution
<img width="1021" alt="Screenshot 2025-03-01 at 12 35 20" src="https://github.com/user-attachments/assets/ef6b9993-3cf0-4b4e-b6d4-b29be92c8111" />

therefore in this pull request I propose to skip adding 1 day in case in "price_to" specifies the time including hours


### Related Pull Requests
<!-- related pull request placeholder -->
--

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#39169

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. add special price to product by POST <host>/rest/<store_code>/V1/products/special-price according [documentation](https://developer.adobe.com/commerce/webapi/rest/modules/catalog/catalog-pricing/)
2. check product price, it will have special price one day longer

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-39169

engcom-Hotel commented:
@magento run all tests

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Thanks for the contribution!

Please fix the failed static test and also please add an automated test in accordance to the [DOD](https://developer.adobe.com/commerce/contributor/guides/code-contributions/definition-of-done/).

Thanks

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- changing Copyright

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Please also add some automated tests.

Thanks

Den4ik commented on lib/internal/Magento/Framework/Stdlib/DateTime/Timezone.php:
```diff
@@ -274,7 +274,7 @@ public function isScopeDateInInterval($scope, $dateFrom = null, $dateTo = null)
         $fromTimeStamp = strtotime($dateFrom);
         $toTimeStamp = strtotime($dateTo);
-        if ($dateTo) {
+        if ($dateTo && date('H:i:s', strtotime($dateTo)) === '00:00:00') {
```
We will have the same +1 day result if use '00:00:00' in the API query. I'm not sure that it's correct solution

KrasnoshchokBohdan commented:
@Den4ik 

> We will have the same +1 day result if use '00:00:00' in the API query. I'm not sure that it's correct solution

pls provide more information

if we update special price with 
"price_to": "2025-07-18 00:00:00" we get "2025-07-18 23:59:59" and this is ok by my view 
but if we update with
"price_to": "2025-07-18 17:00:00" we get "2025-07-19 17:00:00" before fix 
and "2025-07-18 17:00:00" after

> your special price will be visible until tomorrow 17
> so instead expire at 2024-09-12 17:00:00 it will expire on 2024-09-13 17:00:00

https://github.com/magento/magento2/issues/39169 from description

how I tested it 

https://app.magento.test/rest/default/V1/products/special-price (app.magento.test replace with your base_url)
body
{
  "prices": [
    {
      "price": 100,
      "store_id": 0,
      "price_from": "2025-04-08 00:00:00",
      "price_to": "2025-07-18 17:00:00",
      "sku": "test-simple"  // your sku
    }
  ]
}
(date price_from should be not in future)

if we check with POST <host>/rest/<store_code>/V1/products/special-price-information
or in catalog_product_entity_datetime price_to date will be the same as in body (for example "2025-07-18 17:00:00")

but we should triger \Magento\Framework\Stdlib\DateTime\Timezone::isScopeDateInInterval
for this I'm using pdp page (with clean cache)

before fix
<img width="925" alt="image" src="https://github.com/user-attachments/assets/4e630106-289c-45be-aca3-ddeec1db1e05" />

after fix 
<img width="985" alt="image" src="https://github.com/user-attachments/assets/1901443a-ffaf-4c65-a5e2-689031059a31" />

Den4ik commented:
@KrasnoshchokBohdan I mean `+86400` will add 24h to the request date.
> "price_to": "2025-07-18 00:00:00" we get "2025-07-18 23:59:59" and this is ok by my view

After the fix `"price_to": "2025-07-18 00:00:00"` will be `"price_to": "2025-07-19 00:00:00"`

KrasnoshchokBohdan commented:
> After the fix "price_to": "2025-07-18 00:00:00" will be "price_to": "2025-07-19 00:00:00"
but it was the same before fix 

fix is only cofer 

> your special price will be visible until tomorrow 17
> so instead expire at 2024-09-12 17:00:00 it will expire on 2024-09-13 17:00:00

https://github.com/magento/magento2/issues/39169 from description

{
  "prices": [
    {
      "price": 100,
      "store_id": 0,
      "price_from": "2025-04-08 00:00:00",
      "price_to": "2025-07-18 00:00:00",
      "sku": "test-simple"
    }
  ]
}

no changes
<img width="860" alt="image" src="https://github.com/user-attachments/assets/ee73f173-7665-4e91-bba1-ccf619759748" />

after fix
<img width="970" alt="image" src="https://github.com/user-attachments/assets/404c36e4-06fb-4d4e-a860-563f0ca95213" />

sorry it seems I still don't understand the question

KrasnoshchokBohdan commented:
@Den4ik 

> I haven't checked the source code more deeply, but I don't rule out that somewhere the time value may be thrown off and date become 2025-07-19

maybe you meen - date become 2025-07-17?
because, according to documentations 

https://experienceleague.adobe.com/en/docs/commerce-admin/catalog/products/pricing/product-price-special

> The special price goes into effect immediately after midnight at the beginning of the start date (00:01) and continues until just before midnight (23:59) on the day before the end date.

and I dont see any manipulation with special price date except the method we looking 
\Magento\Catalog\Pricing\Price\SpecialPrice::getValue
<img width="937" alt="image" src="https://github.com/user-attachments/assets/b413f7cc-0d62-4a5e-9117-47455b9f73c1" />

so maybe I should replace 
```
// fix date YYYY-MM-DD 00:00:00 to YYYY-MM-DD 23:59:59
            $toTimeStamp += 86400;
```
to 
```
// fix date YYYY-MM-DD 00:00:00 to YYYY-MM-DD 23:59:59
            $toTimeStamp -= 1;
```

<img width="979" alt="image" src="https://github.com/user-attachments/assets/e3332eb5-d1e6-4bf6-a70f-685d7aa6d2e2" />

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- added unit tests

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- fixes for static and unit test

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- fix for static test

KrasnoshchokBohdan commented:
@Den4ik Hi, so should I change ?

```
if ($dateTo && date('H:i:s', strtotime($dateTo)) === '00:00:00') {
    // fix date YYYY-MM-DD 00:00:00 to YYYY-MM-DD 23:59:59
     $toTimeStamp += 86400;
}
```
to 

```
if ($dateTo && date('H:i:s', strtotime($dateTo)) === '00:00:00') {
    // fix date YYYY-MM-DD 00:00:00 to YYYY-MM-DD 23:59:59
     $toTimeStamp += 86399;
}
```

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan commented:
@magento run Integration Tests

Den4ik commented:
> @Den4ik Hi, so should I change ?
> 
> ```
> if ($dateTo && date('H:i:s', strtotime($dateTo)) === '00:00:00') {
>     // fix date YYYY-MM-DD 00:00:00 to YYYY-MM-DD 23:59:59
>      $toTimeStamp += 86400;
> }
> ```
> 
> to
> 
> ```
> if ($dateTo && date('H:i:s', strtotime($dateTo)) === '00:00:00') {
>     // fix date YYYY-MM-DD 00:00:00 to YYYY-MM-DD 23:59:59
>      $toTimeStamp += 86399;
> }
> ```

Hi @KrasnoshchokBohdan 
Sorry for delay with response. I believe yes, we should use `86399`

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan commented:
@magento run Functional Tests B2B

KrasnoshchokBohdan commented:
@magento run Functional Tests B2B,Functional Tests CE

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

engcom-Hotel commented:
As you are still working on this PR, we are moving this into `Changes Requested` bucket.

Thanks

KrasnoshchokBohdan commented:
@magento run Integration Tests

KrasnoshchokBohdan commented:
On my view, failed tests not releted to this pull request (WebAPI and Integration was green before branch merge)

Den4ik reviewed (approved):
@KrasnoshchokBohdan Thanks for update.
Changes looks good for me.
Failed tests looks like unrelated to changes in PR. I'll rerun it to be sure that they are different on each tests execution.

Den4ik reviewed (changes requested):
Sorry, @KrasnoshchokBohdan could you please check WebAPI tests, because they are related to updates

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan commented:
@magento run Integration Tests

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- more targeted solution to cover only the question described in the issue

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan commented:
@magento run Integration Tests

engcom-Charlie commented:
HI @KrasnoshchokBohdan,

Thank you for your contribution!

As I can see that, you are working on the review comments and tests failures, also marked the PR as draft hence I am moving it to on Hold status. Please feel free to Open it once you are done with implementation.

engcom-Charlie commented:
It seems that bot moved it to Review bucket, moving it back to On Hold

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- cover additional isScopeDateInInterval method usage for special price interval checking

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- fix for tests

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan commented:
@magento run Functional Tests B2B, Integration Tests, Magento Health Index

KrasnoshchokBohdan commented:
@Den4ik @engcom-Hotel 
I encountered an issue where tests were breaking in scopes that don’t apply to custom prices, because the method \Magento\Framework\Stdlib\DateTime\Timezone::isScopeDateInInterval is used by several different classes. To address this, I tried another approach that, in my opinion, should only apply to custom prices. I would like to know your opinion on this.

engcom-Hotel commented:
Hello @KrasnoshchokBohdan,

After reviewing the changes in the PR, it appears to offer a reasonable solution to the described issue by limiting the scope to custom prices and addressing the incorrect handling of the time component in `specialPriceTo`.

However, the duplication of logic (extract the `$specialPriceTo` adjustment logic into a reusable method or utility class) and the absence of automated tests should be addressed to enhance the robustness of the solution.

@Den4ik, looking forward to your thoughts on this.

Thanks

engcom-Hotel commented:
Hello @KrasnoshchokBohdan,

Is this PR is ready for review?

KrasnoshchokBohdan commented:
Hello @engcom-Hotel! Thanks for the review and comment. I'll create new service class and unit test for it.

KrasnoshchokBohdan committed:
magento/magento2#39146: Image resize queue error due to .tmp file handling when deleting unsaved product images

- add new service class with unit test

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run all tests

Den4ik commented:
Hi @KrasnoshchokBohdan,
Thanks fro update. Please ping when PR will be ready fro review.

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- fix unit tests

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- fix unit and static tests

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@engcom-Hotel @Den4ik Hi, please take a look at the latest changes, I look forward to your feedback.

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-39169

engcom-Hotel commented:
@magento run all tests

Den4ik commented:
Hi @KrasnoshchokBohdan 
Thanks for updates.
As I mentioned earlier looks like we will have wrong result if I specify `00:00:00` in the request.
For example, I send request with `price_from: 2025-05-11 00:00:00; price_to: 2025-05-12 00:00:00`
In my opinion, expected result should affect only 1 full day 2025-05-11, but in fact it affect 2 days. On the other hand this logic introduced in https://github.com/magento/magento2/commit/c6856aa94c8f354b32629c0f894a55c708deb3b0 11 years ago starting from first Magento release.
So, I'm ok with the fix for specific time case (for example, `17:00:00`)
Lets wait tests result

Den4ik commented on app/code/Magento/Catalog/Model/Product/Type/Price.php:
```diff
@@ -88,6 +89,11 @@ class Price implements ResetAfterRequestInterface
+    /**
+     * @var SpecialPriceService|null
+     */
+    protected ?SpecialPriceService $specialPriceService;
```
Should be private

KrasnoshchokBohdan replied:
thanks for the comment, I would like to clarify, I use it in the child class \Magento\Bundle\Model\Product\Price so it is not private

Den4ik replied:
It will require Core Team approval.
cc @engcom-Hotel

engcom-Hotel replied:
Hello @Den4ik, 

Thanks for highlighting this point.

Hello @KrasnoshchokBohdan,

Is it possible to make it `private` and if access in subclasses is needed, expose a protected getter as below:

```
protected function getSpecialPriceService(): ?SpecialPriceService
{
    return $this->specialPriceService;
}
```

Den4ik commented on app/code/Magento/Catalog/Pricing/Price/SpecialPrice.php:
```diff
@@ -21,29 +23,38 @@ class SpecialPrice extends AbstractPrice implements SpecialPriceInterface, BaseP
+    /**
+     * @var SpecialPriceService
+     */
+    protected SpecialPriceService $specialPriceService;
```
Should be private

Den4ik commented on app/code/Magento/Catalog/Service/SpecialPriceService.php:
```diff
@@ -0,0 +1,38 @@
+ */
+declare(strict_types=1);
+
+namespace Magento\Catalog\Service;
```
namespace `Magento\Catalog\Service` is obsolete. Please move to different namespace to pass static tests

Den4ik reviewed (changes requested):
@KrasnoshchokBohdan Please check my comments

engcom-Hotel commented:
@magento run Database Compare

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- moving SpecialPriceService to different namespace to pass static tests,changing SpecialPriceService property visibility to private

KrasnoshchokBohdan commented:
@magento run all tests

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-39169

engcom-Hotel commented:
@magento run all tests

Den4ik reviewed (approved):
As I mentioned changes looks good for me.
Only one pointt that require attention is failed unit test.
@engcom-Hotel Do we need internal team approval for protected property?

Den4ik reviewed (changes requested):
@KrasnoshchokBohdan Could you please check failed unit test

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Please refer the review comment below.

Thanks

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- adding getter for SpecialPriceService

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan commented:
@Den4ik I can't find this test probably it's only for EE and B2B, sorry
Magento\SalesRuleStaging\Test\Unit\Model\Plugin\DateResolverPluginTest::testBeforeGetFromDate
![image](https://github.com/user-attachments/assets/5849c90d-99f3-4e78-ad17-15f7c12c971f)

Den4ik commented:
@magento run Static Tests

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Thanks for making the changes. Just one small thing, static test is still failing. Request you to please fix that. Other failures seems flaky to me.

Thanks

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

KrasnoshchokBohdan commented:
@magento run Database Compare, Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Magento Health Index, Sample Data Tests B2B, Sample Data Tests CE, Sample Data Tests EE, Semantic Version Checker, Unit Tests, WebAPI Tests

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- revert const PRICE_CODE changes

KrasnoshchokBohdan commented:
@engcom-Hotel Hi, I can't reproduce this issue in my local environment, but according to the screenshot, I assume that the problem is related to changes in \Magento\Catalog\Pricing\Price\SpecialPrice::PRICE_CODE. Therefore, I have rolled back those changes. (I previously changed this constant due to an error in static tests.)

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

It seems you have fixed the mentioned issue by @engcom-Bravo, but we can see some static tests are failing. 

Please fix those as well.

Thanks

KrasnoshchokBohdan committed:
magento/magento2#39169: Special Price To Date is wrongly validated on applySpecialPrice

- changed visibility for const PRICE_CODE

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39169

engcom-Dash committed:
Merge branch '2.4-develop' into fix-for-issue-39169

engcom-Dash commented:
@magento run all tests

engcom-Dash commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

engcom-Dash commented:
The consistent test failures for **Functional B2B** are known Issues and JIRA is raised for them. Other failures are inconsistent and seems to be flaky. They neither part of PR nor failing because of the PR changes.

Build 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39690/8f42d3566719b9958325376679173f3c/Functional/allure-report-b2b/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/610ab46d-917e-4323-a120-1cd561e03243)

Build 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39690/287a336c87e5d0b284d3f8d478a1f240/Functional/allure-report-b2b/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/902204aa-8241-4d67-922f-7cf4fad54202)

**Known Issues:**
OrderWithMultipleConfigurableProductsAdditionalStockTest : [ACQE-8060](https://jira.corp.adobe.com/browse/ACQE-8060)
EnablePaypalExpressCheckoutAndSubmitAnOrderUsingPaypalExpressCheckoutTest : [ACQE-8007](https://jira.corp.adobe.com/browse/ACQE-8007)
StoreFrontSimpleProductWithSpecialAndTierDiscountPriceTest : [ACQE-7971](https://jira.corp.adobe.com/browse/ACQE-7971)

The consistent test failure for **Functional CE** is known Issues and JIRA is raised for the same. Other failures are inconsistent and seems to be flaky. They neither part of PR nor failing because of the PR changes.

Build 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39690/e517e2b946161599488a7d675914bbf4/Functional/allure-report-ce/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/ae1e71f3-1744-4136-916a-783163e5346a)

Build 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39690/b6ec0d558b51716982ef4f365b19bb1c/Functional/allure-report-ce/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/181a7324-27b9-4d6e-93ae-a62926de0f43)

**Known Issues:**
EnablePaypalExpressCheckoutAndSubmitAnOrderUsingPaypalExpressCheckoutTest : [ACQE-8007](https://jira.corp.adobe.com/browse/ACQE-8007)

The consistent test failure for **Functional EE** is known Issues and JIRA is raised for the same. Other failure is inconsistent and seems to be flaky. They neither part of PR nor failing because of the PR changes.

Build 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39690/3049098b7d976dbc2eb8751f3e56e025/Functional/allure-report-ee/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/bcca1645-d5c6-4e97-aaba-92a61c5668e5)

Build 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39690/7de1f6d980021ed16813b17760fb9d7d/Functional/allure-report-ee/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

![image](https://github.com/user-attachments/assets/8ae57f17-47a4-4ca1-b6db-d1ea9e953133)

**Known Issues:**
EnablePaypalExpressCheckoutAndSubmitAnOrderUsingPaypalExpressCheckoutTest : [ACQE-8007](https://jira.corp.adobe.com/browse/ACQE-8007)

hence moving this PR in Merge In Progress

engcom-Charlie committed:
Merge branch '2.4-develop' into fix-for-issue-39169
